### PR TITLE
Allow Option/Optional key/value serde in Consumer/Producer Settings #211 #212

### PR DESCRIPTION
**Pull request by glammers1**
This PR is the result of the discussion in #213. The old one does not apply at all.

* Now APIs (Consumer and Producer) allows `apply` with `Option` parameters and `create` with `Optional` parameters. If they are not passed to settings will check configuration and only then throw illegal argument exception.

```scala
def apply[K, V](
      system: ActorSystem,
      keyDeserializer: Option[Deserializer[K]],
      valueDeserializer: Option[Deserializer[V]])

def apply[K, V](
      config: Config,
      keyDeserializer: Option[Deserializer[K]],
      valueDeserializer: Option[Deserializer[V]])

 def apply[K, V](
      system: ActorSystem,
      keyDeserializer: Deserializer[K],
      valueDeserializer: Deserializer[V])

def apply[K, V](
      config: Config,
      keyDeserializer: Deserializer[K],
      valueDeserializer: Deserializer[V])
```
**Similar 4 for java**

* Remove the other variant that has no serializer params.

* Additional two new tests were added for both Consumer/Producer Settings. One of these tests for the Producer reproduce the validation error explained in #211.

* Change the `ProducerSettings` API to resolve the bug.

### CODE REVIEW

**Comment by kciesielski** [comment]
LGTM

File: core/src/main/scala/akka/kafka/ConsumerSettings.scala [hunk 1]
```diff
@@ -172,6 +153,7 @@ object ConsumerSettings {
   /**
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/ConsumerSettings.scala]: regarding above `apply(system, Some(keyDeserializer), Some(valueDeserializer))`
I think we should add
```
require(keyDeserializer != null, "keyDeserializer must be defined")
require(valueDeserializer != null, "valueDeserializer must be defined")
```

Otherwise user can by mistake pass `null` and we end up with `Some(null)`

Comment (glammers1) [inline core/src/main/scala/akka/kafka/ConsumerSettings.scala]: Yes, we could add guard for null values. What is the best way? 

1.Change `Some` to `Option`.

`apply(system, Option(keyDeserializer), Option(valueDeserializer))`

null values would be `None` and it would be validated in the other `apply` method. We have nothing more to change.

2.Add specific validation as you comment.

I vote for 1 because for `create` methods that also calls `apply` using `scala-java8-compat`:

`apply(config, keySerializer.asScala, valueSerializer.asScala)`

`asScala` is converting `null` values to `None`.

**Comment by glammers1** [comment]
PR updated!

* Now validations check `null` values in the main `apply` method (others always calls this one).

```scala
require(
      keySerializer != null &&
        (keySerializer.isDefined || properties.contains(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG)),
      "Key serializer should be defined or declared in configuration"
    )
```

```scala
require(
      valueSerializer != null &&
        (valueSerializer.isDefined || properties.contains(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG)),
      "Value serializer should be defined or declared in configuration"
    )
```

* Change `Some` to `Option` in the overloads of `apply` in order to obtain the same behavior of `scala-java8-compat` `asScala` in `create`.

* Add two new tests to both `ProducerSettingsSpec` and `ConsumerSettingsSpec` to check all the possible calls with `null` values:

```scala
ConsumerSettings(conf, new ByteArrayDeserializer, null)
``` 
uses the following `apply`

```scala
def apply[K, V](
    config: Config,
    keyDeserializer: Deserializer[K],
    valueDeserializer: Deserializer[V]
  ): ConsumerSettings[K, V] = {
    apply(config, Option(keyDeserializer), Option(valueDeserializer))
  }
```

And

```scala
ConsumerSettings(conf, None, null)
```

uses the following `apply`

```scala
def apply[K, V](
    config: Config,
    keyDeserializer: Option[Deserializer[K]],
    valueDeserializer: Option[Deserializer[V]]
  ): ConsumerSettings[K, V] = { ... }
```

Same for `ProducerSettings`

IMHO, all the cases are covered.

**Comment by patriknw** [comment]
LGTM, thanks
